torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
   return torch::createPyObject(atensor);
 }
 
+// In cases like data loader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with multiprocessing.queue.,
+// will hang waiting for data. This is difficult to avoid on PyTorch side as it
+// can be caused by limited shm, other libraries users call in the workers. The
+// following methods is an effort to do our best provide some error message to
+// users when such unfortunate events happen.
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::mutex worker_pid_mutex;
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static pid_t *worker_pids;
+static size_t num_worker_pids = 0;
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+    signal(SIGBUS, handler_SIGBUS);
+    signal(SIGSEGV, handler_SIGSEGV);
+    Py_RETURN_NONE;
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
+  int status;
+  pid_t p;
+  pid_t *pid_ptr;
+
+  // The flags and status checks ensure that we are really observing a child
+  // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+  // https://stackoverflow.com/a/40707100
+  while ((p = waitpid(-1, &status, WNOHANG|WUNTRACED|WCONTINUED)) > 0) {
```

Comment:
This will fail on windows, right?...

Reply:
Most of this patch is likely to fail on windows I think. I don't think there's a similar signalling mechanism there. @peterjc123 do you have any hints for making this portable?

Reply:
I'm also a bit worried of this use of `waitpid` and a custom signal handler for SIGCHLD. It might e.g. overwrite Python's handlers if it uses them to get return codes in `multiprocessing`. Also, if you use `waitpid` without `NOWAIT`, it will reap the child, so Python won't be able to retrieve the exit code.

Reply:
A reference from the implementation from Apache [here](https://reviews.apache.org/r/46340/diff/2#index_header). They just defined a Windows version of `os.waitpid`, supporting the option arg `WNOHANG`. But i don't think there is sth like that in Windows. A better way to implement it may be using threads or processes like what i stated below.